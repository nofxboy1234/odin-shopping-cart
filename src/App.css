.nav-link {
  background-color: white;
  color: black;
  cursor: pointer;
}

.nav-link:hover {
  background-color: black;
  color: white;
}

.product {
  display: flex;
  flex-direction: column;
  padding: 1rem;
  border: 1px solid black;
  border-radius: 5px;
  margin: 1rem;
  width: 250px;
  height: 300px;
}

.product:hover {
  box-shadow: 5px 5px 5px rgb(112, 112, 112);
}

.content {
  margin: 1rem;
}
