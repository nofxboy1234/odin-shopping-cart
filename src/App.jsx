import Navigation from './components/navigation/Navigation';
import Content from './components/content/Content';
import { useState } from 'react';
import './App.css';
import { Outlet } from 'react-router-dom';

const tempProducts = [
  {
    id: 1,
    imageUrl: '/products/1',
    title: 'product 1',
    price: 10.0,
  },
  {
    id: 2,
    imageUrl: '/products/2',
    title: 'product 2',
    price: 20.5,
  },
  {
    id: 3,
    imageUrl: '/products/3',
    title: 'product 3',
    price: 139.99,
  },
];

function App() {
  const [products, setProducts] = useState(tempProducts);
  const [cart, setCart] = useState([]);

  const getProductById = (itemId) =>
    products.find((product) => product.id === itemId);

  const getItemInCartById = (productId) =>
    cart.find((item) => item.id === productId);

  const addItemToCart = (newItem) => {
    const cartItem = getItemInCartById(newItem.id);
    if (cartItem) {
      newItem.quantity += cartItem.quantity;
      updateItemInCart(newItem);
    } else {
      setCart([...cart, newItem]);
    }

    return newItem.quantity;
  };

  const updateItemInCart = (editedItem) => {
    const updatedItems = cart.map((item) => {
      if (editedItem.id === item.id) {
        return editedItem;
      } else {
        return item;
      }
    });

    setCart(updatedItems);
  };

  const removeItemFromCart = (itemToRemoveId) => {
    const updatedItems = cart.filter((item) => item.id != itemToRemoveId);
    setCart(updatedItems);
  };

  console.log('rendering App');

  return (
    <>
      <Navigation
        cartCount={cart.reduce(
          (total, cartItem) => total + cartItem.quantity,
          0
        )}
      />
      <Outlet
        context={{
          cart,
          products,
          addItemToCart,
          updateItemInCart,
          getProductById,
          removeItemFromCart,
        }}
      />
    </>
  );
}

export default App;
