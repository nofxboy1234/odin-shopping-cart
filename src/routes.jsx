import App from './App';
import Home from './components/home/Home';
import Shop from './components/shop/Shop';
import Food from './components/food/Food';
import Drinks from './components/drinks/Drinks';
import Cart from './components/cart/Cart';
import ErrorPage from './components/errorPage/ErrorPage';

const routes = [
  {
    path: '/',
    element: <App />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: 'shop',
        element: <Shop />,
        children: [
          {
            path: 'food',
            element: <Food />,
          },
          {
            path: 'drinks',
            element: <Drinks />,
          },
        ],
      },
      {
        path: 'cart',
        element: <Cart />,
      },
      {
        index: true,
        element: <Home />,
      },
    ],
  },
];

export default routes;
