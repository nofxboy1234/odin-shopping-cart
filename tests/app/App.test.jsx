import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import App from '../../src/App';
import routes from '../../src/routes/routes';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';

describe('App component', () => {
  it('renders the loading message', () => {
    render(<App />);
    const paragraph = screen.getByRole('paragraph');
    expect(paragraph.textContent).toMatch(/Loading.../i);
  });

  it('renders the Home link', async () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/'],
      initialIndex: 0,
    });

    render(<RouterProvider router={router} />);

    const link = await screen.findByRole('link', { name: 'Home' });
    expect(link).toBeInTheDocument();
  });

  describe.skip('renders the navigation bar', () => {});
});
