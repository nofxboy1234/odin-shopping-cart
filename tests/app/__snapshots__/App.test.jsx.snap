// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > when on the root page > renders the navigation bar above the home page 1`] = `
<div>
  <nav
    class="_navigation_56c144"
  >
    <a
      class="_navLink_56c144"
      href="/"
    >
      Home
    </a>
    <a
      class="_navLink_56c144"
      href="/shop"
    >
      Shop
    </a>
    <a
      class="_navLink_56c144"
      href="/cart"
    >
      Cart (
      0
      )
    </a>
  </nav>
  <h1
    class="_heading_4012cb"
  >
    Welcome to Shopping Cart!
  </h1>
</div>
`;
